cmd/drand-gossip-relay/lp2p/ctor.go

```diff
@@ -0,0 +1,133 @@
+package lp2p
+
+import (
+	"context"
+	"crypto/rand"
+	"fmt"
+	"io/ioutil"
+	mrand "math/rand"
+	"os"
+	"time"
+
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/namespace"
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/libp2p/go-libp2p"
+	"github.com/libp2p/go-libp2p-core/crypto"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-peerstore/pstoreds"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	pubsubpb "github.com/libp2p/go-libp2p-pubsub/pb"
+	libp2ptls "github.com/libp2p/go-libp2p-tls"
+	ma "github.com/multiformats/go-multiaddr"
+	"golang.org/x/crypto/blake2b"
+	xerrors "golang.org/x/xerrors"
+)
+
+var (
+	log      = logging.Logger("lp2p")
+	privFile = "identity.key"
+)
+
+func PubSubTopic(nn string) string {
+	return fmt.Sprintf("/drand/pubsub/v0.0.0/%s", nn)
+}
+
+func ConstructHost(ds datastore.Datastore, priv crypto.PrivKey, listenAddr string,
+	bootstrap []ma.Multiaddr) (host.Host, *pubsub.PubSub, error) {
+	ctx := context.Background()
+
+	pstoreDs := namespace.Wrap(ds, datastore.NewKey("/peerstore"))
+	pstore, err := pstoreds.NewPeerstore(ctx, pstoreDs, pstoreds.DefaultOpts())
+	if err != nil {
+		return nil, nil, xerrors.Errorf("creating peerstore: %w", err)
+	}
+	peerId, err := peer.IDFromPrivateKey(priv)
+	if err != nil {
+		return nil, nil, xerrors.Errorf("computing peerid: %w", err)
+	}
+	err = pstore.AddPrivKey(peerId, priv)
+	if err != nil {
+		return nil, nil, xerrors.Errorf("adding priv to keystore: %w", err)
+	}
+
+	addrInfos, err := peer.AddrInfosFromP2pAddrs(bootstrap...)
+	if err != nil {
+		fmt.Printf("%+v", bootstrap)
+		return nil, nil, xerrors.Errorf("parsing addrInfos: %+v", err)
+	}
+
+	h, err := libp2p.New(ctx,
+		libp2p.ListenAddrStrings(listenAddr),
+		libp2p.Identity(priv),
+		libp2p.Security(libp2ptls.ID, libp2ptls.New),
+		libp2p.DisableRelay(),
+		libp2p.Peerstore(pstore),
+	)
+	if err != nil {
+		return nil, nil, xerrors.Errorf("constructing host: %w", err)
+	}
+
+	p, err := pubsub.NewGossipSub(ctx, h,
+		pubsub.WithPeerExchange(true),
+		pubsub.WithMessageIdFn(func(pmsg *pubsubpb.Message) string {
+			hash := blake2b.Sum256(pmsg.Data)
+			return string(hash[:])
+		}),
+		pubsub.WithDirectPeers(addrInfos),
+		pubsub.WithFloodPublish(true),
+	)
+	if err != nil {
+		return nil, nil, xerrors.Errorf("constructing pubsub: %d", err)
+	}
+
+	go func() {
+		mrand.Shuffle(len(addrInfos), func(i, j int) {
+			addrInfos[i], addrInfos[j] = addrInfos[j], addrInfos[i]
+		})
+		for _, ai := range addrInfos {
+			ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+			err := h.Connect(ctx, ai)
```

Comment:
Connect will return fairly synchronously in general. is that the intention here? If you want to do the connect attempts sequentially, you want to wait for the host event of connection succeeding / failing.

Reply:
I want to ensure libp2p tries connecting to these hosts. Which afaik is what `Connect` does, it blocks until there is a connection or returns right away if there is a connection already.